lib/std/unicode.zig

```diff
@@ -62,32 +112,35 @@ pub fn utf8Encode(c: u21, out: []u8) !u3 {
     return length;
 }
 
-const Utf8DecodeError = Utf8Decode2Error || Utf8Decode3Error || Utf8Decode4Error;
-
-/// Decodes the UTF-8 codepoint encoded in the given slice of bytes.
-/// bytes.len must be equal to utf8ByteSequenceLength(bytes[0]) catch unreachable.
-/// If you already know the length at comptime, you can call one of
-/// utf8Decode2,utf8Decode3,utf8Decode4 directly instead of this function.
-pub fn utf8Decode(bytes: []const u8) Utf8DecodeError!u21 {
-    return switch (bytes.len) {
-        1 => @as(u21, bytes[0]),
-        2 => utf8Decode2(bytes),
-        3 => utf8Decode3(bytes),
-        4 => utf8Decode4(bytes),
+/// Decodes the UTF-8 codepoint encoded in the given slice of bytes and returns
+/// then length of the character decoded.
+///
+/// Guaranteed to not read bytes past this character.
+///
+/// "ret" cannot be *u21 because when casting to *u32 it would have differn't
+/// behavior on Little-Endian and Big-Endian machines, which is too much to ask
+/// of our callers.
+pub fn utf8Decode(bytes: []const u8, ret: *align(4) u32) !u3 {
+    var len = try utf8ByteSequenceLength(bytes[0]);
+    if (bytes.len < len) {
+        return error.Utf8ShortChar;
+    }
+    ret.* = switch (len) {
+        1 => @as(u32, bytes[0]),
+        2 => try utf8Decode2(bytes[0..2]),
+        3 => try utf8Decode3(bytes[0..3]),
+        4 => try utf8Decode4(bytes[0..4]),
         else => unreachable,
     };
+    return len;
 }
 
-const Utf8Decode2Error = error{
-    Utf8ExpectedContinuation,
-    Utf8OverlongEncoding,
-};
-pub fn utf8Decode2(bytes: []const u8) Utf8Decode2Error!u21 {
+pub fn utf8Decode2(bytes: []const u8) !u32 {
     assert(bytes.len == 2);
-    assert(bytes[0] & 0b11100000 == 0b11000000);
```

Comment:
Was this assert incorrect?

Reply:
the new assert is identical

Reply:
oh right; that took me a bit to understand.

Why the new one? my intuition is that it would be slower to execute.